### PR TITLE
XDP re-architecture

### PR DESCRIPTION
**Pull request by hackwa**
This change makes xdp library code more extensible. 
The code has been cleaned up to get rid of circular dependencies and made more modularized.

### CODE REVIEW

File: src/runtime_src/CMakeLists.txt [hunk 1]
```diff
@@ -61,6 +61,7 @@ set_target_properties(xilinxopencl PROPERTIES VERSION ${XRT_VERSION_STRING}
   SOVERSION ${XRT_SOVERSION})
 
 target_link_libraries(oclxdp
+  xdp
```
Comment (stsoe) [inline src/runtime_src/CMakeLists.txt]: doesn't xdp have to be added to dependencies for oclxdp some where around line 24?

Comment (hackwa) [reply src/runtime_src/CMakeLists.txt]: Yep good point Soren. I'll add that.

**Comment by stsoe** [comment]
ok to test

**Comment by hackwa** [comment]
retest this please.

**Comment by hackwa** [comment]
retest this please

**Comment by stsoe** [comment]
retest this please

**Comment by hackwa** [comment]
One of the tests seems to be failing as it's not doing clReleaseprogram:
WARNING: Profiling may contain incomplete information. Please ensure all OpenCL objects are released by your host code (e.g., clReleaseProgram()).

Looking into why it returns non zero.

**Pull request merged**